feat: reuse signed request when reading state

lmuntaner opened the pull request:
# Description

Enable the `readState` of the Agent to reuse the request.

## Motivation

Signing with the hardware wallet requires the user to accept from the device. Reusing the signed request will improve the user experience for hardware wallet users.

# How Has This Been Tested?

I linked a branch of nns-dapp to the changes in this PR and validated that everything worked, not just the new feature.

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

lmuntaner commented:
@krpeacock could you review these changes?

I would love to remove the `any` type. Yet, the transform returns `unknwon` so not sure how I can go around that. Do you have any idea?

krpeacock commented on packages/agent/src/agent/http/index.ts:
```diff
@@ -386,8 +385,19 @@ export class HttpAgent implements Agent {
 
     // Apply transform for identity.
-    transformedRequest = await id?.transformRequest(transformedRequest);
+    return id?.transformRequest(transformedRequest);
```
@lmuntaner is removing this `await` intentional?

lmuntaner replied:
No, just because I am returning the value, I don't think it's needed.

If we had a try/catch then yes, but since we don't, then I don't think so.

krpeacock commented:
@lmuntaner I think the approach here is good, and I see the usefulness for hardware wallets where each request requires another prompt. 

It's unfortunate to expand the use of `any`, but you're using it in the same way it's already being used, so I'm not going to make you refactor the script to get this feature in. 

I'm ready to approve, if we add a new e2e test and document the new feature in the changelog

lmuntaner commented:
Yes, I agree the use of `any` is very ugly. I haven't found a better way unfortunately.

The test and docs sound good, when I have some time I will get to it. Thanks.

lmuntaner commented:
@krpeacock ready for another review

I added an e2e tests to read_state to make sure it also works when the request is passed. I also added a few points to the changelog.

Let me know what you think.

lmuntaner commented:
@krpeacock is the e2e test and the docs that I added what you needed??

krpeacock reviewed (approved):
Looks good, thanks for the work on this one!
